### PR TITLE
[REF] stock: different mtso

### PR DESCRIPTION
**Pull request by mama-odoo**
Currently the mts else mto rule only works if the total quantity is available in stock. However in a majority of case it's not useful and we would like to complete only the missing quantity.

The idea is to always set the initial move as MTS and create a new procurement with the missing part.

Ohter solution would be to split the move in 2 parts, one with the MTO link and the other as take from stock. However this solution could create confusion for users that would not understand the split due to procure_method (invisible field)

Third solution would let the action_assign take from quants even if the origin moves are not fulfill.

We prefer first solution since the second would add complexity for the user and last one will break the behavior for people that want a stong link between picking/SO/MO/PO.

upgrade : coming

### CODE REVIEW

File: addons/stock/models/stock_move.py [hunk 1]
```diff
@@ -1314,27 +1314,33 @@ def _action_confirm(self, merge=True, merge_into=False):
...
 
-        move_create_proc, move_to_confirm, move_waiting = self.browse(move_create_proc), self.browse(move_to_confirm), self.browse(move_waiting)
+        move_create_proc, move_waiting = self.browse(move_create_proc), self.browse(move_waiting)
 
         # create procurements for make to order moves
         procurement_requests = []
-        for move in move_create_proc:
+        procurement_quantities = move_create_proc._prepare_procurement_qty()
+        for move, quantity in zip(move_create_proc, procurement_quantities):
+            if quantity == 0:
```
Comment (amoyaux) [inline addons/stock/models/stock_move.py:1331]: float_is_sero

File: addons/stock/models/stock_move.py [hunk 2]
```diff
@@ -1652,7 +1700,7 @@ def _action_assign(self, force_qty=False):
             else:
                 if float_is_zero(move.product_uom_qty, precision_rounding=move.product_uom.rounding) and not force_qty:
                     assigned_moves_ids.add(move.id)
-                elif not move.move_orig_ids:
+                elif not move.move_orig_ids or (move.rule_id and move.rule_id.procure_method == 'mts_else_mto'):  # and not move_line_ids):
```
Comment (amoyaux) [inline addons/stock/models/stock_move.py:1703]: probably to remove

File: addons/stock/models/stock_move.py [hunk 3]
```diff
@@ -1382,27 +1382,38 @@ def _action_confirm(self, merge=True, merge_into=False):
...
                 key = (move.group_id.id, move.location_id.id, move.location_dest_id.id)
                 to_assign[key].add(move.id)
 
-        move_create_proc, move_to_confirm, move_waiting = self.browse(move_create_proc), self.browse(move_to_confirm), self.browse(move_waiting)
+        move_create_proc, move_waiting = self.browse(move_create_proc), self.browse(move_waiting)
 
         # create procurements for make to order moves
         procurement_requests = []
-        for move in move_create_proc:
+        procurement_quantities = move_create_proc._prepare_procurement_qty()
```
Comment (amoyaux) [inline addons/stock/models/stock_move.py:1397]: Could be cleaned and use only one function _prepapre_procurement

File: addons/stock/models/stock_move.py [hunk 4]
```diff
@@ -1382,27 +1382,38 @@ def _action_confirm(self, merge=True, merge_into=False):
             if move.move_orig_ids:
                 move_waiting.add(move.id)
             else:
-                if move.procure_method == 'make_to_order':
+                if move.procure_method == 'make_to_order' or (move.rule_id and move.rule_id.procure_method == 'mts_else_mto'):
                     move_create_proc.add(move.id)
                 else:
                     move_to_confirm.add(move.id)
```
[the remark this answers is not in the document]
Comment (mama-odoo) [reply addons/stock/models/stock_move.py:1388]: N.B. : selon le nouveau fonctionnement `move.rule_id.procure_method == 'mts_else_mto'` ⇒ `move.procure_method == 'make_to_stock'` , il n'est donc pas nécessaire de vérifier `ou MTSO` pour la confirmation

**Comment by wtaferner** [comment]
Just a comment to be notified about the progress.

Not sure if it is already covered but I kinda expect that expected incoming and outgoing moves and dates are taken into consideration based on the planned date of the procurement for the decision at confirmation of the use case triggering.

Keep up the spirit and make me happy in 18.0 🤣

File: addons/stock/models/stock_move.py [hunk 5]
```diff
@@ -2287,3 +2313,14 @@ def _get_product_catalog_lines_data(self, parent_record=False, **kwargs):
...
             'readOnly': False,
         }
+
+    def _is_mtso(self):
+        self.ensure_one()
+        return self.procure_method == 'make_to_stock' and \
+               self.rule_id and self.rule_id.procure_method == 'mts_else_mto'
+
+    def _get_next_moves(self):
+        return self.move_dest_ids | self.group_id.group_dest_ids.stock_move_ids
```
Comment (amoyaux) [inline addons/stock/models/stock_move.py:2323]: filter out product not related

**Comment by mama-odoo** [comment]
Replaced by :  odoo/odoo#177185

**Pull request closed without merging**